beluga/include/beluga/sensor/likelihood_3d_field_model.hpp

```diff
@@ -0,0 +1,167 @@
+ * Probabilistic Robotics \cite thrun2005probabilistic, Chapter 6.4,
+ * for further reference.
+ *
+ * \note This class satisfies \ref SensorModelPage.
+ *
+ * \tparam OpenVDB grid type.
+ */
+template <typename GridT, typename U>
+class Likelihood3DFieldModel {
+ public:
+  /// State type of a particle.
+  using state_type = Sophus::SE3d;
+
+  /// Weight type of the particle.
+  using weight_type = double;
+
+  /// Measurement type given by the interface.
+  using measurement_type = U;
+
+  /// Map representation type.
+  using map_type = GridT;
+
+  /// Parameter type that the constructor uses to configure the likelihood field model.
+  using param_type = Likelihood3DFieldModelParam;
+
+  /// Constructs a Likelihood3DFieldModel instance.
+  /**
+   * \param params Parameters to configure this instance.
+   *  See beluga::Likelihood3DFieldModelParam for details.
+   * \param grid Narrow band Level set grid representing the static map that the sensor model
+   *  uses to compute a likelihood field for lidar hits and compute importance weights
+   *  for particle states.
+   *  Currently only supports OpenVDB Level sets.
+   */
+  explicit Likelihood3DFieldModel(const param_type& params, const map_type& grid)
+      : params_{params},
+        looker_{openvdb::tools::ClosestSurfacePoint<map_type>::create(grid)},
+        squared_sigma_{params.sigma_hit * params.sigma_hit},
+        amplitude_{params.z_hit / (params.sigma_hit * std::sqrt(2 * Sophus::Constants<double>::pi()))},
+        offset_{params.z_random / params.max_laser_distance} {
+    /// Pre-computed parameters
+    assert(squared_sigma_ > 0.0);
+    assert(amplitude_ > 0.0);
+    // Remember this OpenVDB DualGridSmpler
+  }
+
+  /// Returns a state weighting function conditioned on 3D lidar hits.
+  /**
+   * \param measurement 3D lidar measurement containing the hit points and the transform to the origin.
+   * \return a state weighting function satisfying \ref StateWeightingFunctionPage
+   *  and borrowing a reference to this sensor model (and thus their lifetime are bound).
+   */
+  [[nodiscard]] auto operator()(measurement_type&& measurement) const {
+    const size_t pointcloud_size = measurement.points().size();
+    // Transform each point from the sensor frame to the origin frame
+    auto transformed_points =
+        ranges::views::all(measurement.points()) | ranges::views::transform([&](const auto& point) {
+          return measurement.origin() * point.template cast<double>();
+        });
+
+    return [this, pointcloud_size, points = std::move(transformed_points)](const state_type& state) -> weight_type {
+      std::vector<float> nb_distances;
+      std::vector<openvdb::Vec3R> vdb_points;
+      vdb_points.resize(pointcloud_size);
+
+      // Transform each point to every particle state
+      std::transform(
+          points.begin(), points.end(), vdb_points.begin(), [this, pointcloud_size, state](const auto& point) {
```

Comment:
@pvela2017 why do you need `pointcloud_size` here?

Reply:
Changed in 015f1b79441f103c2e55d662d87fa4c144e0ef2c